File: app/src/playstoreRelease/java/com/renard/ocr/analytics/AnalyticsWithGoogle.java
package com.renard.ocr.analytics;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.renard.ocr.cropimage.image_processing.BlurDetectionResult;
import com.renard.ocr.documents.creation.ocr.LayoutQuestionDialog;
import com.renard.ocr.main_menu.language.OcrLanguage;

/**
 * @author renard
 */
class AnalyticsWithGoogle implements Analytics {
    static final String TRACKING_PREF_KEY = "tracking_key";


    public static final String CATEGORY_OCR = "Ocr";
    public static final String CATEGORY_LANGUAGE = "Language";
    public static final String CATEGORY_DOCUMENT_OPTIONS = "Document Options";
    private static final String LOG_TAG = AnalyticsWithGoogle.class.getSimpleName();
    public static final String START_TTS_SELECTED = "Start Tts selected";
    public static final String SHARE_TEXT_SELECTED = "Share text selected";
    public static final String COPY_TEXT_SELECTED = "Copy text selected";
    public static final String CREATE_PDF_SELECTED = "Create Pdf selected";
    public static final String ACTION_RESULT_DIALOG_SHOWN = "Result Dialog shown";
    public static final String SHOW_TIPS_SELECTED = "Show tips selected";
    public static final String SEND_FEEDBACK_SELECTED = "Send feedback selected";
    public static final String ACTION_BLUR_DIALOG = "Blur Dialog shown";
    public static final String CATEGORY_TEXT_TO_SPEECH = "Text to speech";
    private static final String CATEGORY_INSTALL = "Install";
    public static final String CATEGORY_SCAN_ACCURACY = "Scan accuracy";
    private final Context mApplicationContext;
    private final FirebaseAnalytics firebaseAnalytics;


    public AnalyticsWithGoogle(Context applicationContext) {
        mApplicationContext = applicationContext;
        firebaseAnalytics = FirebaseAnalytics.getInstance(applicationContext);
        listenForGoogleAnalyticsOptOut();
    }

    private void listenForGoogleAnalyticsOptOut() {
        SharedPreferences userPrefs = PreferenceManager.getDefaultSharedPreferences(mApplicationContext);
        userPrefs.registerOnSharedPreferenceChangeListener((sharedPreferences, key) -> {
            if (key.equals(TRACKING_PREF_KEY)) {
                final boolean optOut = sharedPreferences.getBoolean(key, false);
                Log.i(LOG_TAG, "tracking preference was changed. setting app opt out to: " + optOut);
                firebaseAnalytics.setAnalyticsCollectionEnabled(optOut);
            }
        });
    }

    @Override
    public boolean getAppOptOut() {
        SharedPreferences userPrefs = PreferenceManager.getDefaultSharedPreferences(mApplicationContext);
        return userPrefs.getBoolean(TRACKING_PREF_KEY, false);
    }

    @Override
    public void toggleTracking(boolean optOut) {
        Log.i(LOG_TAG, "toggleTracking(" + optOut + ")");
        SharedPreferences userPrefs = PreferenceManager.getDefaultSharedPreferences(mApplicationContext);
        userPrefs.edit().putBoolean(TRACKING_PREF_KEY, optOut).apply();
    }

    @Override
    public void sendScreenView(String screenName) {
        Log.i(LOG_TAG, "Setting screen name: " + screenName);
        Bundle bundle = new Bundle();
        bundle.putString(FirebaseAnalytics.Param.SCREEN_NAME, screenName);
    }


    @Override
    public void sendStartDownload(OcrLanguage language) {
        sendEvent(CATEGORY_LANGUAGE, "Start download", language.getValue(), 1);
    }

    @Override
    public void sendDeleteLanguage(OcrLanguage language) {
        sendEvent(CATEGORY_LANGUAGE, "Delete language", language.getValue(), 1);
    }

    @Override
    public void sendOcrResult(String language, int accuracy) {
        int rem = accuracy % 10;
        final int bracket = accuracy - rem;
        sendEvent(CATEGORY_SCAN_ACCURACY, language, bracket + "", accuracy);
        sendEvent(CATEGORY_OCR, "Scan completed", language, accuracy);
    }

    @Override
    public void sendOcrLanguageChanged(OcrLanguage language) {
        sendEvent(CATEGORY_OCR, "Picked Language", language.getValue(), 1);
    }

    @Override
    public void sendLayoutDialogCancelled() {
        sendEvent(CATEGORY_OCR, "Scan Cancelled", "Layout Dialog", 1);
    }

    @Override
    public void sendOcrCancelled() {
        sendEvent(CATEGORY_OCR, "Scan Cancelled", "Back Button", 1);
    }

    @Override
    public void sendOcrStarted(String language, LayoutQuestionDialog.LayoutKind layout) {
        sendEvent(CATEGORY_OCR, "Scan started", language, 1);
        sendEvent(CATEGORY_OCR, "Layout chosen", layout.name(), 1);
    }

    @Override
    public void optionTranslateText() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, "Translate started", "", 1);
    }

    @Override
    public void noLanguageInstalled() {
        firebaseAnalytics.logEvent("no_Language_installed",null);
    }


    @Override
    public void optionDocumentViewMode(boolean showingText) {
        String label = showingText ? "Text" : "Image";
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, "View mode changed", label, 1);
    }

    @Override
    public void optionTextSettings() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, "Text settings selected", "", 1);

    }

    @Override
    public void optionTableOfContents() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, "Table of contents selected", "", 1);
    }

    @Override
    public void optionsDeleteDocument() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, "Delete document selected", "", 1);
    }

    @Override
    public void optionsCreatePdf() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, CREATE_PDF_SELECTED, "", 1);
    }

    @Override
    public void optionsCopyToClipboard() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, COPY_TEXT_SELECTED, "", 1);
    }

    @Override
    public void optionsStartTts() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, START_TTS_SELECTED, "", 1);
    }

    @Override
    public void optionsShareText() {
        sendEvent(CATEGORY_DOCUMENT_OPTIONS, SHARE_TEXT_SELECTED, "", 1);
    }

    @Override
    public void ttsLanguageChanged(OcrLanguage lang) {
        sendEvent(CATEGORY_TEXT_TO_SPEECH, "Language changed", lang.getValue(), 1);
    }

    @Override
    public void ttsStart(String language) {
        sendEvent(CATEGORY_TEXT_TO_SPEECH, "Started speaking", language, 1);
    }

    @Override
    public void ttsStop() {
        sendEvent(CATEGORY_TEXT_TO_SPEECH, "Stopped speaking", "", 1);
    }

    @Override
    public void startGallery() {
        sendEvent(CATEGORY_OCR, "Gallery started", "", 1);
    }

    @Override
    public void startCamera() {
        sendEvent(CATEGORY_OCR, "Camera started", "", 1);
    }

    @Override
    public void sendCropError() {
        sendEvent(CATEGORY_OCR, "Crop error occurred", "", 1);
    }

    @Override
    public void sendBlurResult(BlurDetectionResult blurriness) {
        final int blurValue = (int) (blurriness.getBlurValue() * 100);
        sendEvent(CATEGORY_OCR, "Blur test completed", blurriness.getBlurriness().name(), blurValue);
    }

    @Override
    public void newImageBecauseOfBlurWarning(float blurriness) {
        sendEvent(CATEGORY_OCR, ACTION_BLUR_DIALOG, "New image requested", (int) (blurriness * 100));

    }

    @Override
    public void continueDespiteOfBlurWarning(float blurriness) {
        sendEvent(CATEGORY_OCR, ACTION_BLUR_DIALOG, "Continue with blurry image", (int) (blurriness * 100));
    }

    @Override
    public void ocrResultSendFeedback() {
        sendEvent(CATEGORY_OCR, ACTION_RESULT_DIALOG_SHOWN, SEND_FEEDBACK_SELECTED, 1);
    }

    @Override
    public void ocrResultStartTts() {
        sendEvent(CATEGORY_OCR, ACTION_RESULT_DIALOG_SHOWN, START_TTS_SELECTED, 1);
    }

    @Override
    public void ocrResultCopyToClipboard() {
        sendEvent(CATEGORY_OCR, ACTION_RESULT_DIALOG_SHOWN, COPY_TEXT_SELECTED, 1);
    }

    @Override
    public void ocrResultShowTips() {
        sendEvent(CATEGORY_OCR, ACTION_RESULT_DIALOG_SHOWN, SHOW_TIPS_SELECTED, 1);
    }

    @Override
    public void ocrResultCreatePdf() {
        sendEvent(CATEGORY_OCR, ACTION_RESULT_DIALOG_SHOWN, CREATE_PDF_SELECTED, 1);
    }

    @Override
    public void ocrResultShareText() {
        sendEvent(CATEGORY_OCR, ACTION_RESULT_DIALOG_SHOWN, SHARE_TEXT_SELECTED, 1);

    }

    private void sendEvent(String category, String action, String label, int value) {
        Log.i(LOG_TAG, "sendEvent(" + category + ", " + action + ", " + label + ", " + value + ")");
        Bundle params = new Bundle();
        params.putString("category", category);
        params.putString("action", action);
        params.putString("label", label);
        params.putInt("value", value);
        firebaseAnalytics.logEvent(category, params);
    }

    @Override
    public void sendClickYoutube() {
        sendEvent(CATEGORY_INSTALL, "Youtube link clicked", "", 1);
    }

    @Override
    public void sendIgnoreMemoryWarning(long availableMegs) {
        sendEvent(CATEGORY_OCR, "Memory Warning", "ignored", (int) availableMegs);
    }

    @Override
    public void sendHeedMemoryWarning(long availableMegs) {
        sendEvent(CATEGORY_OCR, "Memory Warning", "heeded", (int) availableMegs);
    }


}

File: app/src/playstoreRelease/java/com/renard/ocr/analytics/AnalyticsWithGoogle.kt
package com.renard.ocr.analytics

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.util.Log
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.FirebaseAnalytics.Event.SCREEN_VIEW
import com.google.firebase.analytics.FirebaseAnalytics.Param.SCREEN_NAME
import com.google.firebase.analytics.ktx.logEvent
import com.renard.ocr.cropimage.image_processing.BlurDetectionResult
import com.renard.ocr.documents.creation.ocr.LayoutQuestionDialog.LayoutKind
import com.renard.ocr.main_menu.language.OcrLanguage

/**
 * @author renard
 */
internal class AnalyticsWithGoogle(private val mApplicationContext: Context) : Analytics {
    private val firebaseAnalytics: FirebaseAnalytics = FirebaseAnalytics.getInstance(mApplicationContext)

    private fun listenForGoogleAnalyticsOptOut() {
        val userPrefs = PreferenceManager.getDefaultSharedPreferences(mApplicationContext)
        userPrefs.registerOnSharedPreferenceChangeListener { sharedPreferences: SharedPreferences, key: String ->
            if (key == TRACKING_PREF_KEY) {
                val optOut = sharedPreferences.getBoolean(key, false)
                Log.i(LOG_TAG, "tracking preference was changed. setting app opt out to: $optOut")
                firebaseAnalytics.setAnalyticsCollectionEnabled(optOut)
            }
        }
    }

    override fun getAppOptOut(): Boolean {
        val userPrefs = PreferenceManager.getDefaultSharedPreferences(mApplicationContext)
        return userPrefs.getBoolean(TRACKING_PREF_KEY, false)
    }

    override fun toggleTracking(optOut: Boolean) {
        Log.i(LOG_TAG, "toggleTracking($optOut)")
        val userPrefs = PreferenceManager.getDefaultSharedPreferences(mApplicationContext)
        userPrefs.edit().putBoolean(TRACKING_PREF_KEY, optOut).apply()
    }

    override fun sendScreenView(screenName: String) {
        firebaseAnalytics.logEvent(SCREEN_VIEW) { param(SCREEN_NAME, screenName) }
    }

    override fun sendStartDownload(language: OcrLanguage) {
        firebaseAnalytics.logEvent("download_language") {
            param(PARAM_LANGUAGE, language.value)
        }
    }

    override fun sendDeleteLanguage(language: OcrLanguage) {
        firebaseAnalytics.logEvent("delete_language") {
            param(PARAM_LANGUAGE, language.value)
        }
    }

    override fun sendOcrResult(language: String, accuracy: Int) {
        val bracket = accuracy - accuracy % 10
        firebaseAnalytics.logEvent("scan_accuracy") {
            param(PARAM_LANGUAGE, language)
            param(PARAM_ACCURACY_BRACKET, bracket.toString())
        }
        firebaseAnalytics.logEvent("scan_complete") {
            param(PARAM_LANGUAGE, language)
            param(PARAM_ACCURACY, accuracy.toLong())
        }
    }

    override fun sendOcrLanguageChanged(language: OcrLanguage) {
        firebaseAnalytics.logEvent("change_ocr_language") {
            param(PARAM_LANGUAGE, language.value)
        }
    }

    override fun sendLayoutDialogCancelled() {
        firebaseAnalytics.logEvent("scan_cancelled") {
            param(PARAM_OCR_STEP, "layout_dialog")
        }
    }

    override fun sendOcrCancelled() {
        firebaseAnalytics.logEvent("scan_cancelled") {
            param(PARAM_OCR_STEP, "activity_finished")
        }
    }

    override fun sendOcrStarted(language: String, layout: LayoutKind) {
        firebaseAnalytics.logEvent("scan_started") {
            param("layout", layout.name)
            param(PARAM_LANGUAGE, language)
        }
    }

    override fun optionTranslateText() {
        firebaseAnalytics.logEvent("translate_text")
    }

    override fun noLanguageInstalled() {
        firebaseAnalytics.logEvent("no_Language_installed")
    }

    override fun optionDocumentViewMode(showingText: Boolean) {
        firebaseAnalytics.logEvent("switch_document_view_mode") {
            param("mode", if (showingText) "text" else "image")
        }
    }


    override fun optionTextSettings() {
        firebaseAnalytics.logEvent("select_text_settings")
    }

    override fun optionTableOfContents() {
        firebaseAnalytics.logEvent("select_table_of_contents")
    }

    override fun optionsDeleteDocument() {
        firebaseAnalytics.logEvent("delete_document")
    }

    override fun optionsCreatePdf() {
        firebaseAnalytics.logEvent(EVENT_CREATE_PDF) {
            param(PARAM_CONTEXT, "options")
        }
    }

    override fun optionsCopyToClipboard() {
        firebaseAnalytics.logEvent(EVENT_COPY_TEXT_TO_CLIPBOARD) {
            param(PARAM_CONTEXT, "options")
        }
    }

    override fun optionsShareText() {
        firebaseAnalytics.logEvent(EVENT_SHARE_TEXT) {
            param(PARAM_CONTEXT, "options")
        }
    }

    override fun ttsStart(language: String) {
        firebaseAnalytics.logEvent(EVENT_START_TTS) {
            param(PARAM_LANGUAGE, language)
        }
    }

    override fun startGallery() {
        firebaseAnalytics.logEvent("start_gallery")
    }

    override fun startCamera() {
        firebaseAnalytics.logEvent("start_camera")
    }

    override fun sendCropError() {
        firebaseAnalytics.logEvent("crop_error")
    }

    override fun sendBlurResult(blurriness: BlurDetectionResult) {
        firebaseAnalytics.logEvent("blur_result") {
            param("blurriness", (blurriness.blurValue * 100).toLong())
            param("blurriness_bracket", blurriness.blurriness.name)
        }
    }

    override fun newImageBecauseOfBlurWarning(blurriness: Float) {
        firebaseAnalytics.logEvent("discard_blurred_image") {
            param("blurriness", (blurriness * 100).toLong())
        }
    }

    override fun continueDespiteOfBlurWarning(blurriness: Float) {
        firebaseAnalytics.logEvent("continue_with_blurred_image") {
            param("blurriness", (blurriness * 100).toLong())
        }
    }

    override fun ocrResultSendFeedback() {
        firebaseAnalytics.logEvent("send_feedback_after_scan")
    }

    override fun ocrResultCopyToClipboard() {
        firebaseAnalytics.logEvent(EVENT_COPY_TEXT_TO_CLIPBOARD) {
            param(PARAM_CONTEXT, "ocr_result")
        }
    }

    override fun ocrResultShowTips() {
        firebaseAnalytics.logEvent("show_tips")
    }

    override fun ocrResultCreatePdf() {
        firebaseAnalytics.logEvent(EVENT_CREATE_PDF) {
            param(PARAM_CONTEXT, "ocr_result")
        }
    }

    override fun ocrResultShareText() {
        firebaseAnalytics.logEvent(EVENT_SHARE_TEXT) {
            param(PARAM_CONTEXT, "ocr_result")
        }
    }

    override fun sendClickYoutube() {
        firebaseAnalytics.logEvent("youtube_clicked")
    }

    override fun sendIgnoreMemoryWarning(availableMegs: Long) {
        firebaseAnalytics.logEvent("ignore_low_ram") {
            param("free_ram", availableMegs)
        }
    }

    override fun sendHeedMemoryWarning(availableMegs: Long) {
        firebaseAnalytics.logEvent("abort_scan_low_ram") {
            param("free_ram", availableMegs)
        }
    }

    companion object {
        const val EVENT_SHARE_TEXT = "share_text"
        const val EVENT_CREATE_PDF = "create_pdf"
        const val EVENT_COPY_TEXT_TO_CLIPBOARD = "copy_text_to_clipboard"
        const val EVENT_START_TTS = "start_tts"
        const val PARAM_CONTEXT = "context"
        const val PARAM_ACCURACY_BRACKET = "accuracy_bracket"
        const val PARAM_OCR_STEP = "ocr_step"
        const val PARAM_ACCURACY = "accuracy"
        const val PARAM_LANGUAGE = "language"
        const val TRACKING_PREF_KEY = "tracking_key"
        private val LOG_TAG = AnalyticsWithGoogle::class.java.simpleName
    }

    init {
        listenForGoogleAnalyticsOptOut()
    }
}

private fun FirebaseAnalytics.logEvent(event: String) {
    logEvent(event, null)
}
